File: samples/go/tests/json_schema_chat_test.go
package tests

import (
	"testing"

	chat "samples/go/chat"

	"github.com/stretchr/testify/require"
	"go.temporal.io/sdk/converter"
)

// TestJSONSchemaChatRuntime round-trips every chat wire fixture through the
// Temporal default data converter and asserts JSON-equality against the
// canonical fixtures, mirroring the Python and Java suites.
//
// Exception (see json-schema/nullability.md): optional+nullable fields collapse
// in Go. message-full.json carries replyToId: null (optional+nullable), which
// Go collapses on serialize, so it is verified by deserialization + field checks
// rather than exact JSON-equality, matching the Java test. (room-open.json's
// topic is required-nullable, so its explicit null survives the round-trip and
// is checked via JSON-equality.)
func TestJSONSchemaChatRuntime(t *testing.T) {
	dc := converter.GetDefaultDataConverter()

	minimal := roundTripJSONEq[chat.Message](t, dc, "chat", "message-minimal.json")
	require.Equal(t, chat.MessageKindText, minimal.Kind)
	require.Equal(t, "hi", minimal.Body)
	require.Nil(t, minimal.ReplyToID)
	require.Nil(t, minimal.Priority)
	require.Equal(t, int64(0), minimal.PriorityOrDefault())

	// message-full carries replyToId: null (optional+nullable) — deserialization only.
	full := decodeFixture[chat.Message](t, dc, "chat", "message-full.json")
	require.Nil(t, full.ReplyToID)
	require.NotNil(t, full.Priority)
	require.Equal(t, int64(7), *full.Priority)

	room := roundTripJSONEq[chat.Room](t, dc, "chat", "room-open.json")
	require.Equal(t, "r1", room.RoomID)
	require.Nil(t, room.Topic)
	require.Equal(t, []string{"a"}, room.Members)
	require.Contains(t, room.AdditionalProperties, "x-extra")

	labels := roundTripJSONEq[chat.Labels](t, dc, "chat", "labels.json")
	require.Equal(t, "prod", labels.AdditionalProperties["env"])
	require.Equal(t, "core", labels.AdditionalProperties["team"])

	input := roundTripJSONEq[chat.SendMessageInput](t, dc, "chat", "send-message-input.json")
	require.Equal(t, "r1", input.RoomID)
	require.Equal(t, "hi", input.Message.Body)

	output := roundTripJSONEq[chat.SendMessageOutput](t, dc, "chat", "send-message-output.json")
	require.Equal(t, "m1", output.MessageID)
}

File: samples/typescript/tests/json-schema-chat.test.ts
import { describe, expect, test } from "vitest";
import type { TransferTypeConverter } from "nexus-rpc";

import {
  DEFAULT_PRIORITY,
  labelsTransferTypeConverter,
  messageTransferTypeConverter,
  roomTransferTypeConverter,
  sendMessageInputTransferTypeConverter,
  sendMessageOutputTransferTypeConverter,
  ValidationError,
  type Labels,
  type Message,
} from "../chat/index.ts";
import {
  fixtureBytes,
  loadFixture as loadFixtureFrom,
  roundTripFixture,
} from "./json-converter-helper.ts";

const wireFixtureDir = new URL("../../wire/json_schema/chat/", import.meta.url);

function loadFixture(name: string): unknown {
  return loadFixtureFrom(wireFixtureDir, name);
}

// Round-trip a fixture through the Temporal data converter (driven by the
// generated converter) and assert the re-serialized JSON is JSON-equal to the
// fixture. TS converters preserve explicit nulls, so all chat fixtures use exact
// JSON-equality (no optional+nullable collapse — unlike Go).
function expectRoundTrip<T>(name: string, converter: TransferTypeConverter<T>): T {
  const { value, serialized } = roundTripFixture(
    converter,
    fixtureBytes(wireFixtureDir, name),
  );
  expect(serialized).toEqual(loadFixture(name));
  return value;
}

describe("json-schema chat generated definitions", () => {
  test("roundtrips canonical wire fixtures through the Temporal converter", () => {
    const message = expectRoundTrip(
      "message-minimal.json",
      messageTransferTypeConverter,
    );
    expect(message).toMatchObject<Message>({
      kind: "text",
      body: "hi",
    });
    expect(message.replyToId).toBeUndefined();
    expect(message.priority ?? DEFAULT_PRIORITY).toBe(0);

    const fullMessage = expectRoundTrip(
      "message-full.json",
      messageTransferTypeConverter,
    );
    expect(fullMessage.replyToId).toBeNull();
    expect(fullMessage.priority).toBe(7);

    const room = expectRoundTrip("room-open.json", roomTransferTypeConverter);
    expect(room.additionalProperties).toEqual({ "x-extra": 42 });

    const labels = expectRoundTrip("labels.json", labelsTransferTypeConverter);
    expect(labels).toMatchObject<Labels>({
      additionalProperties: { env: "prod", team: "core" },
    });

    const request = expectRoundTrip(
      "send-message-input.json",
      sendMessageInputTransferTypeConverter,
    );
    expect(request.message.body).toBe("hi");

    const response = expectRoundTrip(
      "send-message-output.json",
      sendMessageOutputTransferTypeConverter,
    );
    expect(response.messageId).toBe("m1");
  });

  test("reports JSON schema validation errors", () => {
    expect(() =>
      sendMessageInputTransferTypeConverter.fromTransferType({
        roomId: "r1",
        message: { kind: "text", body: "hi" },
        extra: true,
      }),
    ).toThrow(ValidationError);

    expect(() =>
      messageTransferTypeConverter.fromTransferType({ kind: "image", body: "hi" }),
    ).toThrow(ValidationError);

    expect(() => sendMessageOutputTransferTypeConverter.fromTransferType({})).toThrow(
      ValidationError,
    );

    try {
      messageTransferTypeConverter.fromTransferType({ kind: "image", body: "hi" });
      throw new Error("expected invalid payload to fail");
    } catch (error) {
      expect(error).toBeInstanceOf(ValidationError);
      expect((error as ValidationError).type).toBe("BAD_REQUEST");
    }

    expect(() =>
      roomTransferTypeConverter.toTransferType({
        roomId: "r1",
        displayName: "Room",
        topic: null,
        additionalProperties: { roomId: "shadow" },
      }),
    ).toThrow(/roomId.*collides with declared property/);
  });
});
